integration/metrics_test.go

```diff
@@ -64,3 +72,123 @@ func TestMetrics(t *testing.T) {
+disable_verification = {{.DisableVerification}}
+
+[plugins."io.containerd.snapshotter.v1.soci".blob]
+check_always = true
+
+[debug]
+format = "json"
+level = "debug"
+
+{{.AdditionalConfig}}
+`, struct {
+			DisableVerification bool
+			AdditionalConfig    string
+		}{
+			DisableVerification: disableVerification,
+			AdditionalConfig:    additionalConfig,
+		}))
+	}
+	getSnapshotterConfigYaml := func(disableVerification bool) []byte {
+		return []byte(fmt.Sprintf(`
+metrics_address = "localhost:6060"
+disable_verification = %v
+`, disableVerification))
+	}
+
+	sh, done := newShellWithRegistry(t, regConfig)
+	defer done()
+
+	if err := testutil.WriteFileContents(sh, defaultContainerdConfigPath, getContainerdConfigYaml(false), 0600); err != nil {
+		t.Fatalf("failed to write %v: %v", defaultContainerdConfigPath, err)
+	}
+	if err := testutil.WriteFileContents(sh, defaultSnapshotterConfigPath, getSnapshotterConfigYaml(false), 0600); err != nil {
+		t.Fatalf("failed to write %v: %v", defaultSnapshotterConfigPath, err)
+	}
+
+	testCases := []struct {
+		name                  string
+		image                 string
+		indexDigestFn         func(*shell.Shell, imageInfo) string
+		expectedFallbackCount int
+	}{
+		{
+			name:  "image with all layers having ztocs and no fs.Mount error results in 0 overlay fallback",
+			image: "rabbitmq:latest",
+			indexDigestFn: func(sh *shell.Shell, image imageInfo) string {
+				return optimizeImageWithOpts(sh, image, 1<<22, 0)
+			},
+			expectedFallbackCount: 0,
+		},
+		{
+			name:  "image with some layers not having ztoc and no fs.Mount results in 0 overlay fallback",
+			image: "rabbitmq:latest",
+			indexDigestFn: func(sh *shell.Shell, image imageInfo) string {
+				return optimizeImageWithOpts(sh, image, 1<<22, 10<<20)
+			},
+			expectedFallbackCount: 0,
+		},
+		{
+			name:  "image with fs.Mount errors results in non-zero overlay fallback",
+			image: "rabbitmq:latest",
+			indexDigestFn: func(_ *shell.Shell, _ imageInfo) string {
+				return "dwadwadawdad"
+			},
+			expectedFallbackCount: 10,
+		},
+	}
+
+	checkOverlayFallbackCount := func(output string, expected int) error {
+		lines := strings.Split(output, "\n")
+		for _, line := range lines {
+			if !strings.Contains(line, commonmetrics.FuseMountFailureCount) {
+				continue
+			}
+			var got int
+			_, err := fmt.Sscanf(line, `soci_fs_operation_count{layer="",operation_type="fuse_mount_failure_count"} %d`, &got)
+			if err != nil {
+				return err
+			}
+			if got != expected {
+				return fmt.Errorf("unexpected overlay fallbacks: got %d, expected %d", got, expected)
+			}
+			return nil
+		}
+		if expected != 0 {
+			return fmt.Errorf("expected %d overlay fallbacks but got 0", expected)
+		}
+		return nil
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			rebootContainerd(t, sh, "", "")
+
+			imgInfo := dockerhub(tc.image)
+
+			sh.X("ctr", "i", "pull", imgInfo.ref)
+			indexDigest := tc.indexDigestFn(sh, imgInfo)
```

Comment:
Not sure if it has impact, but `ctr i pull` is called twice for some testcases: one here, one in `optimizeImageWithOpts`

https://github.com/awslabs/soci-snapshotter/blob/2cf2a1607c016406946f79d01130dbfc304ed8e5/integration/push_test.go#L211

Reply:
Hmm, I don't think it has any impact.